### PR TITLE
fix: remove raw answer keys from web-layer test contract (#45)

### LINKED ISSUE #45: Test: Remove raw answer keys from web-layer test contract
**Source:** CodeRabbit review on PR #33
**File:** `apps/web/lib/queries/quiz-report.test.ts` (line 94)

The test currently models raw answer keys (e.g., correct option identifiers) as part of the expected shape. This knowledge should not be part of the web-layer contract — correct answers are stripped by the `get_quiz_questions()` RPC. Update the test fixtures and assertions to reflect only what the web layer is permitted to see.

### PR DESCRIPTION
**Pull request by okpilot**
## Summary

- New `get_report_correct_options` RPC moves correct-answer resolution to Postgres. The TypeScript layer never reads the `correct` boolean from options JSONB.
- Quiz report test fixtures no longer contain `correct: true/false` on options — answer keys are fully absent from the web-layer test contract.
- RPC is session-scoped (ownership + completion check), derives question set from `quiz_session_answers` (no caller-supplied IDs), uses `DISTINCT ON` for consistency with other RPCs, and old insecure overloads are explicitly dropped.
- 5 new tests: RPC error handling, null score fallback, empty-answers RPC skip, sessionId forwarding, correct-field stripping assertion.

## Changes

| File | Change |
|------|--------|
| Migrations 032-036 | New RPC + iterative hardening + overload cleanup |
| `quiz-report.ts` | Use RPC for correctOptionId, explicit options projection, RPC error handling |
| `quiz-report.test.ts` | Remove `correct` from fixtures, add RPC mock, 5 new tests (19 total) |
| `types.ts` | Add `get_report_correct_options` function type |
| `docs/database.md` | Document new RPC |
| `docs/decisions.md` | Update Decision 25 with RPC implementation |
| `docs/security.md` | Update post-session exception with RPC approach |
| `docs/plan.md` | Log migrations 032-035 |

## Test plan

- [x] `pnpm check-types` — clean
- [x] `pnpm test` — 881 tests passing (82 files)
- [x] Security auditor — APPROVED (0 critical, 0 high, 0 medium)
- [ ] CodeRabbit review
- [ ] Red-team E2E specs (tracked in #253)

Closes #45

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * New server-side endpoint to provide correct-answer IDs for completed quiz sessions; client no longer receives raw correctness flags.

* **Security Improvements**
  * Stronger session ownership checks, SECURITY-guarded server calls, and removal of insecure overloads to prevent probing and cross-tenant access.

* **Bug Fixes & Validation**
  * Session-scoped parameter required for answer checks, case-insensitive ID validation, and improved error handling for edge cases.

* **Tests & Documentation**
  * Expanded tests, docs, and patterns for RPC boundaries, hook-splitting, and error-path coverage.

* **UX**
  * Loading skeletons and improved error logging for better observability.

### CODE REVIEW

File: packages/db/migrations/035_report_correct_options_distinct_on.sql [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  -- This SECURITY DEFINER function bypasses RLS — do not remove the guard.
+  RETURN QUERY
+  SELECT DISTINCT ON (sa.question_id)
+    sa.question_id,
+    (opt->>'id')::text AS correct_option_id
+  FROM quiz_session_answers sa
+  JOIN questions q ON q.id = sa.question_id
+  CROSS JOIN LATERAL jsonb_array_elements(q.options) AS opt
+  WHERE sa.session_id = p_session_id
+    AND q.deleted_at IS NULL
```
[the remark this answers is not in the document]
Comment (okpilot) [reply packages/db/migrations/035_report_correct_options_distinct_on.sql]: Fixed in 94dce31. Migration 037 removes \ — soft-deleted questions now correctly appear in historical reports. The session's answers already constrain the question set.

File: supabase/migrations/20260316222651_scope_report_correct_options_to_session.sql [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+  RETURN QUERY
+  SELECT
+    q.id AS question_id,
+    (opt->>'id')::text AS correct_option_id
+  FROM questions q,
+  LATERAL jsonb_array_elements(q.options) AS opt
+  WHERE q.id = ANY(p_question_ids)
+    AND q.deleted_at IS NULL
+    AND (opt->>'correct')::boolean = true;
```
[the remark this answers is not in the document]
Comment (okpilot) [reply supabase/migrations/20260316222651_scope_report_correct_options_to_session.sql]: Acknowledged — this is the iterative hardening sequence. Migration 034 removes `p_question_ids` entirely and derives questions from `quiz_session_answers`, and migration 036 drops this `(uuid, uuid[])` overload. The final registered signature is `(uuid)` only.
